numpy/random/mtrand.pyx

```diff
@@ -3753,14 +3808,21 @@ cdef class RandomState:
 
         # GH10839, ensure double to make tol meaningful
         cov = cov.astype(np.double)
-        (u, s, v) = svd(cov)
+        if mode == 'legacy':
+            (u, s, v) = svd(cov)
+        elif mode == 'fast':
+            l = cholesky(cov)
+        else:
+            s, v  = eigh(cov)
 
-        if check_valid != 'ignore':
+        if check_valid != 'ignore' and mode != 'fast':
             if check_valid != 'warn' and check_valid != 'raise':
                 raise ValueError(
                     "check_valid must equal 'warn', 'raise', or 'ignore'")
-
-            psd = np.allclose(np.dot(v.T * s, v), cov, rtol=tol, atol=tol)
+            if mode == 'legacy':
+                psd = np.allclose(np.dot(v.T * s, v), cov, rtol=tol, atol=tol)
+            else:
+                psd = np.all(s >= 0)
```

Comment:
This needs to be tied to the matrix norm

```
psd = np.all(s[-1] >= -np.spacing(100.) * s[0] * len(s))
```

or something similar.

Reply:
Then you can take the abs later if this passes otherwise you might not want to combat with large negative eigs.

Reply:
This makes sense, thanks for the suggestion.